gcsio/src/main/java/com/google/cloud/hadoop/gcsio/DeleteFolderOperation.java

```diff
@@ -0,0 +1,258 @@
+
+  // Maximum number of times to retry deletes in the case of precondition failures.
+  private static final int MAXIMUM_PRECONDITION_FAILURES_IN_DELETE = 4;
+  private static final ApiErrorExtractor errorExtractor = ApiErrorExtractor.INSTANCE;
+
+  // Error extractor to map APi exception to meaningful ErrorTypes.
+  private static final ErrorTypeExtractor errorTypeExtractor = GrpcErrorTypeExtractor.INSTANCE;
+  private final GoogleCloudStorageOptions storageOptions;
+  private final KeySetView<IOException, Boolean> allExceptions;
+  private final List<FolderInfo> folders;
+  private final BatchExecutor batchExecutor;
+  private final StorageControlClient storageControlClient;
+  private final BlockingQueue<FolderInfo> folderDeleteBlockingQueue;
+  private final ConcurrentHashMap<String, Long> countOfChildren;
+
+  DeleteFolderOperation(
+      List<FolderInfo> folders,
+      GoogleCloudStorageOptions storageOptions,
+      StorageControlClient storageControlClient) {
+    this.folders = folders;
+    this.storageOptions = storageOptions;
+    this.storageControlClient = storageControlClient;
+    this.folderDeleteBlockingQueue = new LinkedBlockingQueue<>(folders.size());
+
+    // threads for parallel delete of folder resources
+    this.batchExecutor = new BatchExecutor(this.storageOptions.getBatchThreads());
+
+    // Gather exceptions to wrap in a composite exception at the end.
+    this.allExceptions = ConcurrentHashMap.newKeySet();
+
+    // Map to store number of children for each parent object
+    this.countOfChildren = new ConcurrentHashMap<>();
+  }
+
+  /** Helper function that performs the deletion process for folder resources */
+  public void performDeleteOperation() throws InterruptedException {
+    int folderSize = this.folders.size();
+    computeChildrenForFolderResource();
+
+    // this will avoid infinite loop when all folders are deleted
+    while (folderSize != 0 && encounteredNoExceptions()) {
+      FolderInfo folderToDelete = getElementFromBlockingQueue();
+      folderSize--;
+
+      // Queue the deletion request
+      queueSingleFolderDelete(folderToDelete, /* attempt */ 1);
+    }
+    batchExecutorShutdown();
+  }
+
+  /** Shutting down batch executor and flushing any remaining requests */
+  private void batchExecutorShutdown() {
+    try {
+      this.batchExecutor.shutdown();
+    } catch (IOException e) {
+      this.allExceptions.add(
+          new IOException(
+              String.format("Error in shutting down batch executor : %s", e.getMessage())));
+    }
+  }
+
+  public boolean encounteredNoExceptions() {
+    return this.allExceptions.isEmpty();
+  }
+
+  public KeySetView<IOException, Boolean> getAllExceptions() {
+    return allExceptions;
+  }
+
+  /** Gets the head from the blocking queue */
+  public FolderInfo getElementFromBlockingQueue() throws InterruptedException {
+    return this.folderDeleteBlockingQueue.poll(1, TimeUnit.MINUTES);
```

Comment:
If this throws InterruptedException, let us make sure that we will logs this since this is not expected, correct?